vcs_version fails when building pex binary in a docker environment

**Describe the bug**

vcs_version targets fail to build when pointed at a docker_environment:
`('ERROR: no version found for', Namespace(root='/u/kilogram/src/loose-cannon', config='pyproject.synthetic.toml', strip_dev=False, command=None))`

The `__run.sh` looks the same regardless of the environment:
```__run.sh
#!/usr/bin/env bash
# This command line should execute the same process as pants did internally.
cd /tmp/pants-sandbox-AGNZdd
env -i  ./setuptools_scm.pex_pex_shim.sh --root /u/kilogram/src/loose-cannon --config pyproject.synthetic.toml
```

I'm inferring from this that setuptools itself is not being run in the docker container.

The relevant portions of BUILD look like:

``` BUILD
__defaults__({
  pex_binary: dict(environment="docker"),
  docker_image: dict(build_platform=["linux/amd64"]),
})

vcs_version(
    name = "_version",
    generate_to="libs/_version.py",
    template='vcs_version = "{version}"',
)

docker_environment(name="docker", platform="linux_x86_64", ...)
pex_binary(..., environment="docker")
```

I observed that when the pex_binary's environment is "docker", the sandbox has broken `.cache` symlink:
```
$ ls -al /tmp/pants-sandbox-AGNZdd/.cache/pex_root
lrwxrwxrwx 1 kilogram kilogram 28 Jul 16 16:35 /tmp/pants-sandbox-AGNZdd/.cache/pex_root -> /pants-named-caches/pex_root
```

This of course succeeds when the pex_binary's environment is overriden to a local_environment:
```
$ ls -al /tmp/pants-sandbox-7uPeuH/.cache/pex_root
lrwxrwxrwx 1 kilogram kilogram 46 Jul 16 16:39 /tmp/pants-sandbox-7uPeuH/.cache/pex_root -> /u/kilogram/.cache/pants/named_caches/pex_root
```

**Pants version**
Which version of Pants are you using?

2.22.0a0

**OS**
Are you encountering the bug on MacOS, Linux, or both?

Linux edi 6.9.6-arch1-1 #1 SMP PREEMPT_DYNAMIC Fri, 21 Jun 2024 19:49:19 +0000 x86_64 GNU/Linux
ldd (GNU libc) 2.39
cp311-cp311-manylinux_2_39_x86_64

**Additional info**
Add any other information about the problem here, such as attachments or links to gists, if relevant.

Yup. It works with the local_linux environment:
```
{apricot} edi [17:20:43] » PANTS_SOURCE=$HOME/src/pants pants run docker/print_version
Pantsd has been turned off via Env.
17:20:47.01 [INFO] Initializing Nailgun pool for 40 processes...
17:20:49.69 [INFO] Initializing Nailgun pool for 40 processes...
17:20:53.92 [INFO] Completed: Run setuptools_scm for libs:_version
17:20:53.99 [INFO] Completed: Building docker image print_version/print_version:0.1-813dcd3feb41cd76cad0e12e3e1264c1cc5d2def371cb6113388daac305165cb
PEX=/app/print_version.pex
BinaryVersion(release_tag=None, next_version='0.1', dev_distance='5', revision_hash='629ebb3', dirty_workdir='20240729')
{apricot} edi [17:20:58] » git status
On branch main
Your branch is up to date with 'origin/main'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   pants.toml

no changes added to commit (use "git add" and/or "git commit -a")
{apricot} edi [17:20:59] » git diff
diff --git a/pants.toml b/pants.toml
index 35a19aa..1bcbb8c 100644
--- a/pants.toml
+++ b/pants.toml
@@ -41,7 +41,7 @@ default_repository = "{directory}/{name}"
 extra_env_vars = ["HOME"]

 [environments-preview.names]
-# linux = "//:local_linux"  # <-- Works
-linux = "//:docker"  # <-- breaks
+linux = "//:local_linux"  # <-- Works
+# linux = "//:docker"  # <-- breaks
 local_osx = "//:local_osx"
 docker = "//:docker"
```

The only change is to revert the local changes, and it fails (with the docker environment):
```
edi [17:24:33] » git checkout -- pants.toml
edi [17:24:33] » git status
## main...origin/main
edi [17:24:37] » PANTS_SOURCE=$HOME/src/pants pants run docker/print_version
Pantsd has been turned off via Env.
17:24:39.17 [INFO] Initializing Nailgun pool for 40 processes...
17:24:41.79 [INFO] Initializing Nailgun pool for 40 processes...
17:24:45.78 [ERROR] 1 Exception encountered:

Engine traceback:
  in `run` goal

VCSVersioningError: Trying to determine the version for the libs:_version target at libs:_version, but you are not running in a git worktree.
```

--- a/docs/notes/2.22.x.md
+++ b/docs/notes/2.22.x.md
@@ -121,6 +121,8 @@ The `runtime` field of [`aws_python_lambda_layer`](https://www.pantsbuild.org/2.
 
 `pants export` of a Python resolve will now include generated Python sources (for example, from `protobuf_sources` or `thrift_sources` targets) if the resolve is configured in the new `--export-py-generated-sources-in-resolve` advanced option.
 
+When using the `vcs_version` target, force `setuptools_scm` git operations to run in the local environment, so that the local git state is available to them.
+
 #### Semgrep
 
 The default version of `semgrep` used by the `pants.backends.experimental.tool.semgrep` backend is now version 1.72.0, upgraded from 1.46.0. This version requires Python 3.8 or greater.

--- a/docs/notes/2.23.x.md
+++ b/docs/notes/2.23.x.md
@@ -193,8 +193,6 @@ The docs for the `check` goal have been updated to state that third-party type s
 
 Deprecate the `--export-py-hermetic-scripts` option in favor of the new `--export-py-non-hermetic-scripts-in-resolve` option which allows configuring the hermetic scripts logic on a per-resolve basis.
 
-When using the `vcs_version` target, force `setuptools_scm` git operations to run in the local environment, so that the local git state is available to them.
-
 [The `pants.backend.python.providers.experimental.pyenv` backend](https://www.pantsbuild.org/2.23/reference/subsystems/pyenv-python-provider) now respects the patch version of the interpeter constraints.
 
 The function-as-a-service targets like `python_google_cloud_function`, `python_aws_lambda_function`, and `python_aws_lambda_layer` have had many changes to how they handle platforms:

--- a/src/python/pants/backend/python/util_rules/vcs_versioning.py
+++ b/src/python/pants/backend/python/util_rules/vcs_versioning.py
@@ -64,7 +64,10 @@ async def generate_python_from_setuptools_scm(
     # A MaybeGitWorktree is uncacheable, so this enclosing rule will run every time its result
     # is needed, and the process invocation below caches at session scope, meaning this rule
     # will always return a result based on the current underlying git state.
-    maybe_git_worktree = await Get(MaybeGitWorktree, GitWorktreeRequest())
+    maybe_git_worktree = await Get(
+        MaybeGitWorktree,
+        {GitWorktreeRequest(): GitWorktreeRequest, local_environment_name.val: EnvironmentName},
+    )
     if not maybe_git_worktree.git_worktree:
         raise VCSVersioningError(
             softwrap(
@@ -100,7 +103,13 @@ async def generate_python_from_setuptools_scm(
         ),
     )
 
-    setuptools_scm_pex_get = Get(VenvPex, PexRequest, setuptools_scm.to_pex_request())
+    setuptools_scm_pex_get = Get(
+        VenvPex,
+        {
+            setuptools_scm.to_pex_request(): PexRequest,
+            local_environment_name.val: EnvironmentName,
+        },
+    )
     setuptools_scm_pex, input_digest = await MultiGet(setuptools_scm_pex_get, input_digest_get)
 
     argv = ["--root", str(maybe_git_worktree.git_worktree.worktree), "--config", config_path]

